syntax = "proto3";

package distributed;

import "google/protobuf/timestamp.proto";

message Slice {
  int32 start = 1;
  int32 end = 2;
}

message HelloRequest {
  int32 num_cpus = 1;
}

message HelloResponse {
  int32 id = 1;
}

message HeartbeatRequest {
  int32 id = 1;
  google.protobuf.Timestamp timestamp = 2;
}

message HeartbeatResponse {
  bool ok = 1;
}

message DoneRequest {
  int32 id = 1;
  bytes reward_batch = 2;
}

message DoneResponse {}

message SendStateRequest {
  int32 id = 1;
  bytes state = 2;
}

message SendStateResponse {}

enum ServerEventType {
  UNKNOWN = 0;
  SEND_STATE = 1;
  EVALUATE_BATCH = 2;
  STATE_UPDATE = 3;
  OPTIM_STEP = 4;
}

message SubscribeRequest {
  int32 id = 1;
}

message SubscribeResponse {
  ServerEventType type = 1;

  // only for OPTIM_STEP
  repeated bytes optim_rewards = 2;

  // only for EVALUATE_BATCH
  repeated Slice eval_pop_slices = 3;

  // only for STATE_UPDATE
  optional bytes update_state = 4;
}

service ESService {
  // worker joins
  rpc Hello(HelloRequest) returns (HelloResponse) {}

  // epoch is done
  rpc Done(DoneRequest) returns (DoneResponse) {}

  // worker heartbeat
  rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse) {}

  // send state if server demands it
  rpc SendState(SendStateRequest) returns (SendStateResponse) {}
  
  // subscribe to server events
  rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse) {}
}
